### ISSUE TITLE
Problems with Touchscript Input Module

### ISSUE THREAD

**Issue by TimKraemer** [TouchScript/TouchScript#231]
Hey there,

thanks for the good work! I'm currently facing the problem that the `TouchScript Input Module` doesn't behave the way I expect when using `EventSystem.current.IsPointerOverGameObject()`.

I'm spawning an UI by clicking/tapping on a GameObject. By using `if (!EventSystem.current.IsPointerOverGameObject())` I'm preventing the UI to spawn again (because the click/pan is going through the UI).

There might be different options to solve this (I'm open for suggestions), but I wonder why TouchScripts IsPointerOverGameObject() behaves different (it always returns `false`). Did I miss something?

```c#
    void Update() {
        if (!EventSystem.current.IsPointerOverGameObject()) {
            if ( Input.GetMouseButtonUp(0) ) {
               Debug.Log("valid click - there is no UI here!");
            }
        }
```

**Comment by valyard** [comment]
Hi.

What version are you using?
The IsPointerOverGameObject wasn't implemented (8
I added working code to https://github.com/TouchScript/TouchScript/tree/develop branch.

But how is it that UI doesn't catch touches in your case?
Also, you have to know the internal UI id associated with the touch...

**Comment by TimKraemer** [comment]
I'm using 8.2.
> The IsPointerOverGameObject wasn't implemented (8
Ah that explains that :)

Anyway in the meantime I found a solution that works, but isn't exactly beautiful:
```c#
private void touchesBeganHandler(object sender, TouchEventArgs e) {
  foreach (var point in e.Touches) {
    PointerEventData pointerData = new PointerEventData(EventSystem.current);
    pointerData.position = point.Position;
    List<RaycastResult> results = new List<RaycastResult>();
    EventSystem.current.RaycastAll(pointerData, results);
    if (results.Count > 0 && results[0].gameObject.layer == LayerMask.NameToLayer("UI")) {
      Debug.Log("UI click!"); // user touched an element on "UI" Layer
    }
    else {
      Ray ray = Camera.main.ScreenPointToRay(point.Position);
      RaycastHit hit;
      if ( Physics.Raycast(ray, out hit, 100) ) {
        clicked = hit.transform.gameObject; // this is the 3D gameObject the user touched
      }
    }
}
```

What I'm doing there is 
1) casting a raycast from EventSystem (using the touch-pointer position) and checking if the first hit is in my "UI" Layer - in that case we stop all actions here.
2) In any other case we're casting another raycast from `Camera.main` in the direction of our touch-pointer position and see what 3D object we're hitting.

btw: I had to remove the `UIGesture` Scripts from my UI Components to prevent double activation.

I'd be interested in a more elegant solution (and without having to implement Tap-Gestures for every object I want to tap/click) - but this currently works for me. Thanks!

**Comment by valyard** [comment]
Hm... right, this is kinda ugly. The new TouchScript 9 has a nicer way of using touch.GetOverData() since UI is hit automatically by StandardLayer. Have you tried 9 yet?

**Comment by TimKraemer** [comment]
Looks promising! I'm going to check it out when I find the time :+1:

**Comment by TimKraemer** [comment]
coming back to this... I've finally upgraded to 9 and I like the new way of handling things :+1:
But I still don't get the idea how to handle fullscreen things like camera zoom by two-finger-scale-gesture (`FullscreenLayer` on `Main Camera`) without loosing the ability to tap on UI `Screen Space`-buttons. 
As far as I see it, there is only one or the other (`FullscreenLayer` or `StandardLayer`), right?

**Comment by valyard** [comment]
> As far as I see it, there is only one or the other (FullscreenLayer or StandardLayer), right?

Yes, FullscreenLayer is supposed to be under everything.
StandardLayer combines all previously used layers and can handle UI.

This setup is still puzzling me and I am in the process of reworking it. But without much success so far.

**Comment by TimKraemer** [comment]
once again I had time to focus on this issue. I found a solution with the current 9.0 Alpha3 build. (should work in any 9.0 though):

I solved it by creating a `Fullscreen Layer` in my `MainCamera` and based on the multi-user example I'm using an additional 2D Camera with a `Standard Layer` and a `Canvas` in `Screen Space - Camera` render-mode with UI Camera as render-camera (not sure why this is needed, but it works).

Now I'm able to:
* **drag** and **pinch** my Camera Fullscreen layer to realize camera rotation and zoom
* **tap** on 3D object within my scene
* and **tap** on UI canvas items
